orm/versioning_bucket.go

```diff
@@ -0,0 +1,218 @@
+	for _, m := range matches {
+		var vID VersionedIDRef
+		idData := m.Key[dbKeyLength:]
+		if err := vID.Unmarshal(idData); err != nil {
+			return nil, nil, errors.Wrap(err, "wrong key type")
+		}
+		if vID.Version > highestVersion.Version {
+			highestVersion = vID
+			found = m
+		}
+	}
+	if len(found.Key) == 0 {
+		return nil, nil, errors.Wrap(errors.ErrNotFound, "unknown id")
+	}
+	if tombstone.Equal(found.Value) {
+		return nil, nil, errors.ErrDeleted
+	}
+	obj, err := b.Parse(found.Key, found.Value)
+	if err != nil {
+		return nil, nil, err
+	}
+	return &highestVersion, obj, err
+}
+
+// Get works with a marshalled VersionedIDRef key. Direct usage should be avoided in favour of
+// GetVersion or GetLatestVersion.
+// Unlike the classic Get function it returns:
+//  - ErrNotFound when not found
+//  - ErrDeleted when deleted
+// Object won't be nil in success case
+func (b VersioningBucket) Get(db weave.ReadOnlyKVStore, key []byte) (Object, error) {
+	bz, err := db.Get(b.DBKey(key))
+	switch {
+	case err != nil:
+		return nil, err
+	case bz == nil:
+		return nil, errors.ErrNotFound
+	case tombstone.Equal(bz):
+		return nil, errors.ErrDeleted
+	}
+	return b.Parse(key, bz)
+}
+
+// GetVersion returns the stored object for the given VersionedIDRef.
+// Unlike the classic Get function it returns:
+//  - ErrNotFound when not found
+//  - ErrDeleted when deleted
+// Object won't be nil in success case
+
+func (b VersioningBucket) GetVersion(db weave.ReadOnlyKVStore, ref VersionedIDRef) (Object, error) {
+	key, err := ref.Marshal()
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to unmarshal version id key")
+	}
+	return b.Get(db, key)
+}
+
+// Create stores a the given data. It assigns an ID and initial version number to the object instance and returns the
+// VersionedIDRef which won't be nil on success.
+func (b VersioningBucket) Create(db weave.KVStore, data versionedData) (*VersionedIDRef, error) {
+	if data.GetVersion() != 0 {
+		return nil, errors.Wrap(errors.ErrInput, "version is set on create")
+	}
+	data.SetVersion(1)
+	newID, err := b.idGen.NextVal(db, data)
+	if err != nil {
+		return nil, err
+	}
+	idRef := VersionedIDRef{ID: newID, Version: data.GetVersion()}
+	return b.safeUpdate(db, idRef, data)
+}
+
+// Deprecated: Save will always return an error. Use Create or Update instead.
+func (b VersioningBucket) Save(db weave.KVStore, model Object) error {
+	return errors.Wrap(errors.ErrHuman, "raw save not supported")
+}
+
+// Update persists the given data object with a new derived version key in the storage.
+// The VersionedIDRef returned won't be nil on success and contains the new version number.
+// The currentKey must be the latest one in usage or an ErrDuplicate is returned.
+func (b VersioningBucket) Update(db weave.KVStore, id []byte, data versionedData) (*VersionedIDRef, error) {
+	currentKey := VersionedIDRef{ID: id, Version: data.GetVersion()}
+	// prevent gaps
+	switch exists, err := b.Exists(db, currentKey); {
```

Comment:
why the semi-colon here? `, currentKey);`

Reply:
Well spotted. Switch usually works on single value or none. With this semicolon exists and err is scoped to the switch but no value is defined to be used as source for switching.